### PR TITLE
bugfix for plot for string x values

### LINKED ISSUE #18687: bug: plot two lines, unordered xlabels with type str
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
%matplotlib inline

import pandas as pd
df1 = pd.DataFrame([{'x': '1', 'y': 1}, {'x': '2', 'y': 2}])
df2 = pd.DataFrame([{'x': '2', 'y': 3}, {'x': '1', 'y': 4}])

print(df1)
#    x  y
# 0  1  1
# 1  2  2
print(df2)
#    x  y
# 0  2  3
# 1  1  4

# Wrong plot
ax = None
ax = df1.plot('x', 'y', ax=ax)
ax = df2.plot('x', 'y', ax=ax)

# Correct plot
ax = None
ax = df1.sort_values(by=['x']).plot('x', 'y', ax=ax)
ax = df2.sort_values(by=['x']).plot('x', 'y', ax=ax)
```
#### Problem description
I want to plot multiple dataframes in one graph. The x values are strings. The x value order in both dataframes is different. 

The first plot draws the line
 - `x = ['1', '2']` and `y = [1, 2]`
The second plot draws the line 
 - `x = ['2', '1']` and `y = [4, 3]`
Since the second plot overwrites the xticks, the first line is now `x = ['2', '1']` and `y = [1, 2]`.

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Linux
OS-release: 4.10.0-38-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: de_DE.UTF-8
LOCALE: de_DE.UTF-8

pandas: 0.21.0
pytest: 3.3.0
pip: 9.0.1
setuptools: 36.5.0.post20170921
Cython: 0.27.3
numpy: 1.13.3
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.6.3
patsy: 0.4.1
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.0
openpyxl: 2.4.9
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 0.999999999
sqlalchemy: 1.1.13
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by boeddeker**
- [ ] closes #18687
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Matplotlib can handle string data, so `_get_xticks` does not need to convert strings to int

### CODE REVIEW

File: pandas/plotting/_core.py [hunk 1]
```diff
@@ -573,6 +573,8 @@ def _get_xticks(self, convert_period=False):
                 self.data = self.data[notna(self.data.index)]
                 self.data = self.data.sort_index()
                 x = self.data.index._mpl_repr()
+            elif all(isinstance(val, str) for val in index):
```
Comment (jschendel) [inline pandas/plotting/_core.py]: I think the general preference is to use `pandas.compat.string_types` for `isinstance` checks against strings.  Looks like `string_types` is already imported, so just switching to `isinstance(val, string_types)` should be fine.

Comment (boeddeker) [reply pandas/plotting/_core.py]: Ok, I changed the line

File: pandas/plotting/_core.py [hunk 2]
```diff
@@ -573,6 +573,8 @@ def _get_xticks(self, convert_period=False):
                 self.data = self.data[notna(self.data.index)]
                 self.data = self.data.sort_index()
                 x = self.data.index._mpl_repr()
+            elif all(isinstance(val, string_types) for val in index):
```
Comment (TomAugspurger) [inline pandas/plotting/_core.py]: This will be slow for large indexes. Try `index.is_object()`.

Comment (boeddeker) [reply pandas/plotting/_core.py]: Is `is_object` equal to `is_str`? The reason for the generator is that in my example the datatype of the index was wrong (i.e. object and not str) and the generator is only slow when there are str objects. My assumption was that in the case that the index contains str, the index size is relative small.

Comment (jreback) [reply pandas/plotting/_core.py]: you can do
``index.inferred_type in ['string', 'unicode']`` is pretty performant

**Comment by jreback** [comment]
can you add a whatsnew note in 0.22
can you add a test for this

File: pandas/plotting/_core.py (hunk 2 above)
Comment (boeddeker) [reply pandas/plotting/_core.py]: ok, I changed the code to use `index.inferred_type`

**Comment by boeddeker** [comment]
I add a whatsnew note and a test

File: pandas/tests/plotting/test_frame.py [hunk 3]
```diff
@@ -139,6 +139,23 @@ def test_plot(self):
...
+        x2, y2 = ['b', 'a'], [4, 3]
+        df1 = pd.DataFrame(y1, index=x1)
+        df2 = pd.DataFrame(y2, index=x2)
+        ax = None
+        ax = df1.plot(ax=ax)
+        ax = df2.plot(ax=ax)
+
+        line1, line2 = ax.lines
+        # xdata should not be touched (Earlier it was [0, 1])
+        np.testing.assert_equal(line1.get_xdata(), x1)
```
Comment (jreback) [inline pandas/tests/plotting/test_frame.py]: we don't use ``np.testing``, instead use ``tm.assert_numpy_array_equal``

File: pandas/tests/plotting/test_frame.py [hunk 4]
```diff
@@ -139,6 +139,23 @@ def test_plot(self):
...
+        x1, y1 = ['a', 'b'], [1, 2]
+        x2, y2 = ['b', 'a'], [4, 3]
+        df1 = pd.DataFrame(y1, index=x1)
+        df2 = pd.DataFrame(y2, index=x2)
+        ax = None
+        ax = df1.plot(ax=ax)
+        ax = df2.plot(ax=ax)
+
+        line1, line2 = ax.lines
+        # xdata should not be touched (Earlier it was [0, 1])
```
Comment (jreback) [inline pandas/tests/plotting/test_frame.py]: blank line before comments

File: doc/source/whatsnew/v0.22.0.txt [hunk 5]
```diff
@@ -292,7 +292,7 @@ Plotting
 ^^^^^^^^
 
 - :func: `DataFrame.plot` now raises a ``ValueError`` when the ``x`` or ``y`` argument is improperly formed (:issue:`18671`)
--
+- Bug in :func: `DataFrame.plot` where the ``xtickvalues`` are overwritten (:issue:`#18687`)
 -
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: no space after ``:func:``, can you add a bit more text on what a user would want to know here

issue is: ``(:issue:`18687`)`` (no ``#``)

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.22.0.txt (hunk 5 above)
Comment (boeddeker) [reply doc/source/whatsnew/v0.22.0.txt]: The line above mine has also a whitespace after `:func:`

File: pandas/tests/plotting/test_frame.py (hunk 3 above)
Comment (jreback) [reply pandas/tests/plotting/test_frame.py]: sure it does
turn the list into a numpy array
you can not compare a list and an array in our testing by default

Comment (boeddeker) [reply pandas/tests/plotting/test_frame.py]: ok, I made that change

**Comment by jreback** [comment]
pls rebase

**Comment by boeddeker** [comment]
Ok, I did a rebase with master.
But I was not able to merge my changes in whats new, so I dropped my note there.
```
- Bug in :func:`DataFrame.plot` the second call writing in the same figure where both have ``x`` values of type string overwrote the ``xticklabels`` of the first call (:issue:`18687`)
-
```

**Comment by TomAugspurger** [comment]
Can you add another commit with the release note? Those need to be included in the docs.

It looks like this changes is failing other tests: https://travis-ci.org/pandas-dev/pandas/jobs/340101818#L2288. Do those fail for you locally?

**Comment by boeddeker** [comment]
Where should I add a release note? The file where I originally added the note, has completely changed.

Regarding the test, the test also fails local on my machine.
I think the reason is that there is a bug in matplotlib, because the example below does not return the expected plot. `get_xticklabels` is empty (i.e. all values are Text(0,0,'')), while the plot contains labels.
So maybe a bugfix in matplotlib is recommented, before this PR is merged into pandas.
```python
_, ax = plt.subplots()
ax.plot(x, y)
ax.set_xticks([1, 2])  # -> x labels: ['', 'a', 'b']
print('xticklabels', list(ax.get_xticklabels()))  # xticklabels [Text(0,0,''), Text(0,0,'')]
print('xticks', list(ax.get_xticks()))  # xticks [1.0, 2.0]


_, ax = plt.subplots()
ax.plot(x, y)
ax.set_xticks(['b', 'c'])  # -> x labels: ['', 'a', 'b']
print('xticklabels', list(ax.get_xticklabels()))  # xticklabels [Text(0,0,''), Text(0,0,'')]
print('xticks', list(ax.get_xticks()))  # xticks [1.0, 2.0]

```

**Comment by jreback** [comment]
note goes in whatsnew  for 0.23.0 / bugfixes / plotting

**Comment by boeddeker** [comment]
According to https://github.com/matplotlib/matplotlib/issues/10547 matplotlib will change in version 2.2 (currently it is 2.2.rc1) the categorical code, so in my mind, this PR could only be merged, when pandas set the minimum required version of matplotlib to version 2.2.

I do not know, what pandas convention is for requiring a minimum version of a library, especially in the case of matplotlib (The `install_requires` in setup.py does not contain matplotlib)

**Comment by TomAugspurger** [comment]
Our minimum matplotlib is pretty old. We usually work around bugs.

@boeddeker does your test pass on matplotlib 2.2? I think the best thing to do is get pandas passing on matplotlib 2.2 (https://github.com/pandas-dev/pandas/pull/19702) and skip this test on MPL older than 2.2. 

Could you also post images of the output on MPL 2.1 and MPL 2.2?

**Comment by boeddeker** [comment]
I haven't installed matplotlib 2.2 because my cases work. When Version 2.2 is released, I will investigate how there categorical plot work and further work on this PR.

**Comment by jreback** [comment]
@TomAugspurger should we fix / close this?

**Comment by TomAugspurger** [comment]
We should get this passing and merged.

@boeddeker can you track down why the CI tests failed?

**Comment by boeddeker** [comment]
@TomAugspurger Currently I am busy, but in two weeks I have time. I will take care of this PR in two weeks.

**Comment by jreback** [comment]
closing as stale, if you want to continue working, pls ping and we can re-open. you will need to merge master.

**Pull request closed without merging**